aptos-move/framework/src/natives/cryptography/algebra/mod.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright © Aptos Foundation
+
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    natives::{
+        cryptography::algebra::gas::GasParameters,
+        helpers::{make_safe_native, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safely_pop_arg,
+};
+use aptos_types::on_chain_config::{FeatureFlag, Features, TimedFeatures};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use better_any::{Tid, TidAble};
+use move_core_types::language_storage::TypeTag;
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Value, VectorRef},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{any::Any, collections::VecDeque, hash::Hash, ops::Add, rc::Rc, sync::Arc};
+
+pub mod gas;
+
+/// Equivalent to `std::error::not_implemented(0)` in Move.
+const MOVE_ABORT_CODE_NOT_IMPLEMENTED: u64 = 0x0C0000;
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum Structure {
+    BLS12381Fq12,
+    BLS12381G1Affine,
+    BLS12381G2Affine,
+    BLS12381Gt,
+    BLS12381Fr,
+}
+
+impl Structure {
+    pub fn from_type_tag(type_tag: &TypeTag) -> Option<Structure> {
+        match type_tag.to_string().as_str() {
+            "0x1::algebra_bls12381::Fr" => Some(Structure::BLS12381Fr),
+            "0x1::algebra_bls12381::Fq12" => Some(Structure::BLS12381Fq12),
+            "0x1::algebra_bls12381::G1Affine" => Some(Structure::BLS12381G1Affine),
+            "0x1::algebra_bls12381::G2Affine" => Some(Structure::BLS12381G2Affine),
+            "0x1::algebra_bls12381::Gt" => Some(Structure::BLS12381Gt),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Eq, Hash, PartialEq)]
+pub enum SerializationFormat {
+    BLS12381Fq12LscLscLscLsb,
+    BLS12381G1AffineCompressed,
+    BLS12381G1AffineUncompressed,
+    BLS12381G2AffineCompressed,
+    BLS12381G2AffineUnompressed,
+    BLS12381Gt,
+    BLS12381FrLsb,
+    BLS12381FrMsb,
+}
+
+impl TryFrom<u64> for SerializationFormat {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        match value {
+            0x0400000000000000 => Ok(SerializationFormat::BLS12381Fq12LscLscLscLsb),
+            0x0600000000000000 => Ok(SerializationFormat::BLS12381G1AffineUncompressed),
+            0x0601000000000000 => Ok(SerializationFormat::BLS12381G1AffineCompressed),
+            0x0800000000000000 => Ok(SerializationFormat::BLS12381G2AffineUnompressed),
+            0x0801000000000000 => Ok(SerializationFormat::BLS12381G2AffineCompressed),
+            0x0900000000000000 => Ok(SerializationFormat::BLS12381Gt),
+            0x0A00000000000000 => Ok(SerializationFormat::BLS12381FrLsb),
```

Comment:
This is consistent with what's defined in `bls12381_algebra.move`.

But, like we discussed last time, we do not want to worry about format constant clashing across different Move modules: we'd need to account for the module that the constant is associated with.

But, it doesn't matter, I still propose we switch to type-arguments for `SerializationFormat`'s, which would avoid all this.

Reply:
IMO format-as-a-struct has some minor downside compared with format-as-a-u64:
- When calling `serialize()`, in the format-as-a-struct case, we have to specify both the structure type and the format type, i.e., `serialize<Group1, Format1>(element)`, while in format-as-a-u64 case we can leverage Move type inference: `serialize(format1(), some_element)`.
- With format-as-a-struct, the backend need to be aware of the full path of the format struct (e.g. `0x1::algebra_bls12381::Format1`). Unlikely, but if we rename/move the move structs, or there are other changes on how the full path is constructed, we will also need to update the backend. With format-as-a-u64, we have an independent format-ID system which is less likely to be affected.

Reply:
Regarding your point (2) above, with constants we need to be aware of the Move constant in the Rust backend, so that's a bit worse in my mind (e.g., you cannot `grep` as easily for that constant to match it, compared to a struct typename).

Regarding the changes to the full path (i.e., to what typename gets returned by the Move VM) I don't think that is a concern nor likely to happen because: (a) it will be a breaking change everywhere else in the `aptos-core` Move & Rust code and the `move` repo and (2) it will be a breaking change in our Move module anyway, because `Structure:: from_type_tag` will stop working.

Reply:
Regarding point (1) I agree it's more verbose. When called generically, there's no way to make it shorter. But when calling for specific curves, inline functions can make it less verbose.

I would still prefer verbosity over dealing with awkward constants in Rust & over the awkward syntax with the function call.

Reply:
Plus, consistency would be another plus: we use type arguments for everything in this module.

Reply:
Switched back to format-as-a-struct and it looks good!